src/libexpr/flake/call-flake.nix

```
@@ -12,7 +12,36 @@ overrides:
 
 let
 
-  lockFile = builtins.fromJSON lockFileStr;
+  origLockFile = builtins.fromJSON lockFileStr;
+
+  # Build an optimized lock file by consolidating equivalent inputs.
+  lockFile =
+    let
+      nodeCacheKey = key: node:
+        let taint = if overrides ? ${key} then { taintedAs = key; } else {};
+        in builtins.toJSON (node.locked or {} // taint);
+
+      inputsCache = builtins.listToAttrs (builtins.attrValues
+        (builtins.mapAttrs
+          (key: node: { name = nodeCacheKey key node; value = key; })
+          origLockFile.nodes
+        )
+      );
+
+      resolveInputsFromCache = inputs:
+        builtins.mapAttrs
+          (inputName: inputSpec:
+            let resolvedInputName = resolveInput inputSpec;
+            in inputsCache.${nodeCacheKey resolvedInputName lockFile.nodes.${resolvedInputName}})
+          inputs;
```

Resolving a single input name seems like a good primitive. I'd make the caller responsible for the `mapAttrs`.
The `inputName` is unused and not part of this operation.

```suggestion
      deduplicateInputName = inputName: inputSpec:
            let resolvedInputName = resolveInput inputSpec;
            in inputsCache.${nodeCacheKey resolvedInputName lockFile.nodes.${resolvedInputName}};
```